### PR TITLE
Implementation of min eigenvalue in pure TF using Lanczos

### PR DESCRIPTION
**Pull request by AlexeyKurakin**

### CODE REVIEW

File: cleverhans/experimental/certification/utils.py [hunk 1]
```diff
@@ -244,3 +246,82 @@ def minimum_eigen_vector(x, num_steps, learning_rate, vector_prod_fn):
...
+def tf_lanczos_smallest_eigval(vector_prod_fn,
+                               matrix_dim,
+                               max_iter=1000,
+                               collapse_tol=1e-12,
+                               dtype=tf.float32):
+  """Computes smallest eigenvector and eigenvalue using Lanczos in pure TF.
+
+  Args:
+    vector_prod_fn: function which takes a vector as an input and returns
+      matrix vector product.
```
Comment (npapernot) [inline cleverhans/experimental/certification/utils.py]: Can you write this with some high level math notation or point to https://en.wikipedia.org/wiki/Lanczos_algorithm#The_algorithm so people know what you are talking about

File: cleverhans/experimental/certification/utils.py [hunk 2]
```diff
@@ -244,3 +246,82 @@ def minimum_eigen_vector(x, num_steps, learning_rate, vector_prod_fn):
...
+
+  # alpha will store diagonal elements
+  alpha = tf.TensorArray(dtype, size=1, dynamic_size=True, element_shape=())
+  # beta will store off diagonal elements
+  beta = tf.TensorArray(dtype, size=0, dynamic_size=True, element_shape=())
+  # q will store Krylov space basis
+  q_vectors = tf.TensorArray(
+      dtype, size=1, dynamic_size=True, element_shape=(matrix_dim, 1))
+
+  b = tf.random_normal(shape=(matrix_dim, 1), dtype=dtype)
```
Comment (npapernot) [inline cleverhans/experimental/certification/utils.py]: Document: "Create arbitrary vector with Euclidean norm 1."

File: cleverhans/experimental/certification/utils.py [hunk 3]
```diff
@@ -244,3 +246,82 @@ def minimum_eigen_vector(x, num_steps, learning_rate, vector_prod_fn):
...
+  w = w_
+
+  # Subsequent iterations of Lanczos
+  for i in tf.range(1, max_iter):
+    cur_beta = tf.norm(w)
+    if cur_beta < collapse_tol:
+      # return early if Krylov subspace collapsed
+      break
+
+    w = w / cur_beta
```
Comment (npapernot) [inline cleverhans/experimental/certification/utils.py]: is it possible for `cur_beta` to be 0?

Comment (AlexeyKurakin) [reply cleverhans/experimental/certification/utils.py]: cur_beta > collapse_tol at this point. Added comment explaining this

**Pull request merged**